src/Auth.ts

```diff
@@ -339,206 +194,430 @@ export default class Auth {
+          break;
+        }
+        await this.clientStorage.remove({ key });
+        break;
+      case "expo-secure-storage":
+        if (typeof this.clientStorage.deleteItemAsync !== "function") {
+          console.error(`this.clientStorage.deleteItemAsync is not a function`);
+          break;
+        }
+        this.clientStorage.deleteItemAsync(key);
+        break;
+      default:
+        break;
+    }
+  }
+
+  private _generateHeaders(): null | types.Headers {
+    if (this.useCookies) return null;
+
+    return {
+      Authorization: `Bearer ${this.currentSession.getSession()?.jwt_token}`,
+    };
+  }
+
+  private _autoLogin(refreshToken: string | null): void {
+    if (this.ssr) {
+      this._clearSession();
+      return;
+    }
+
+    this._refreshToken(refreshToken);
+  }
+
+  private async _refreshToken(initRefreshToken?: string | null): Promise<void> {
+    const refreshToken =
+      initRefreshToken || (await this._getItem("nhostRefreshToken"));
+
+    if (!refreshToken) {
+      // place at end of call-stack to let frontend get `null` first (to match SSR)
+      setTimeout(() => {
+        this._clearSession();
+      }, 0);
+
+      this.loading = false;
+      return;
+    }
+
+    let res;
+    try {
+      // set lock to avoid two refresh token request being sent at the same time with the same token.
+      // If so, the last request will fail because the first request used the refresh token
+      if (this.refreshTokenLock) {
+        return console.debug(
+          "refresh token already in transit. Halting this request."
+        );
+      }
+      this.refreshTokenLock = true;
+
+      // make refresh token request
+      res = await this.httpClient.get("/token/refresh", {
+        params: {
+          refresh_token: refreshToken,
+        },
+      });
+    } catch (error) {
+      if (error.response?.status === 401) {
+        await this.logout();
+        return
+      } else {
+        return; // silent fail
+      }
+    } finally {
+      // release lock
+      this.refreshTokenLock = false;
+    }
+
+    this._setSession(res.data);
+    this.loading = false;
+    this.tokenChanged();
+  }
+
+  private tokenChanged(): void {
+    for (const tokenChangedFunction of this.tokenChangedFunctions) {
+      tokenChangedFunction();
+    }
+  }
+
+  private authStateChanged(state: boolean): void {
+    for (const authChangedFunction of this.authChangedFunctions) {
+      authChangedFunction(state);
+    }
+  }
+
+  private async _clearSession(): Promise<void> {
+    clearInterval(this.refreshInterval);
+    clearInterval(this.refreshSleepCheckInterval);
 
-    // set refresh token
-    if (!this.use_cookies) {
-      await this.setItem("refresh_token", res.data.refresh_token);
+    this.currentSession.clearSession();
+    this._removeItem("nhostRefreshToken");
+  }
+
+  private async _setSession(session: types.Session) {
+    this.currentSession.setSession(session);
```

Comment:
early exit if no sessions exist, such as after register() when auto-activate user is `false`.

Reply:
@elitan what do you mean by `no sessions exist`? `session` here cannot be `null`.

Reply:
if the setting "auto activate account on regsitration" is `false` and you register a new user the `session` will be `null` during registration.

Reply:
@elitan you're right, I missed those cases